### DISCUSSION TITLE
Interfacing multiple DS18b20

### DISCUSSION THREAD

**Discussion by CodingGhost** [zephyrproject-rtos/zephyr#55743] [Q&A] [answered]
I am wondering how to Interface multiple ds18b20 by using the sensor API.
I already have one sensor working, and I think I could get multiple to work by putting more sensors in the devicetree, but the problem is, that I wont know which is which, as I dont have the ID (ROM address) of the sensors.

Is there a way to Interface multiple ds18b20 in a way that lets me permanently identify them?
Another thing to think of, might be the situation where you dont know how many sensors are connected. In my case, I know there is a max of 3 sensors though so it should not be a problem.
@mbolivar 
@str4t0m

**Comment by CodingGhost** [comment]
any info/opinions about this?

**Comment by str4t0m** [comment]
Yes interfacing multiple ds18b20 is possible. And putting multiple sensors in the device tree will allow you to handle the w1 sensors like any other sensors during runtime. It is also used as an indication to the 1-wire subsystem to use a match rom command instead of the skip rom command in order to select a specific slave.

However due to the fact that 1-wire slaves can only be identified via a unique rom-id, an additional setup step is required where you assign a rom-id to a device. For the ds18b20 this can be done via the sensor attribute interface.
You might have a look at [w1_sensor.h](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/sensor/w1_sensor.h).

>  I dont have the ID (ROM address) of the sensors.

You can use `w1_search_rom()` to list all devices on the bus, as demonstrated in the `drivers/w1/scanner` sample.

> Is there a way to Interface multiple ds18b20 in a way that lets me permanently identify them?

This duty really depends on the use case and is therefore the responsibility of the application.
For example, in order to be able to address a specific sensor(i.e. front sensor, rear sensor), you could execute a routine once, where you assign a rom to a specific "object". Store that assignment on a permanent memory (i.e. nvs, settings). Your application then can assign the roms during program initialization.

> the situation where you dont know how many sensors are connected

Yeah this would be a valid use case, and you could detect them via a search. However if such slaves don't have programmed any identifier onto them it would be hard to find out where they are connected and what they measure.

**Comment by CodingGhost** [comment]
thanks for your answer!
so in my case, as I have 1-3 sensors connected, I put 3 sensors in the device tree, correct?
then I can use w1_search_rom to find out how many sensors are currently connected and what rom they have.
but how do I get the device now? Previously I did DEVICE_DT_GET_ANY(maxim_ds18b20);
but thats not correct now I guess.
I dont quite unserstand what w1_rom_to_sensor_value does either to be honest. Do I use that function instead of the normal sensor_sample_fetch?
also, as googling this stuff, especially the w1_rom_to_sensor_value function spits out about nothing, maybe it would be a good idea to document this scenario somewhere, maybe even with a little example. (Or did I just not find it?)

**Comment by CodingGhost** [comment]
@str4t0m I tried to understand how todo it, but I cant figure it out... I can search and get the ROMs of all connected devices, and I think I understand what w1_rom_to_sensor_value does, but I dont know why I would want to do this..?
I somehow need to get a struct device that I can use with sensor_channel_get, so I would want a w1_rom_to_device function or not?

**Comment by CodingGhost** [comment]
@str4t0m any more Infos on this?

**Comment by CodingGhost** [comment, marked as answer]
I just got it working.
just some infos for the next guy who stumbles across this:
make sure to have at least 2 devices in the devicetree like this:
```
w1_0: w1-zephyr-serial-0 {
		 compatible = "zephyr,w1-serial";
		 #address-cells = <1>;
		 #size-cells = <0>;
		 status = "okay";
 
		 testlabel: ds18b20 {
			 compatible = "maxim,ds18b20";
			 family-code = <0x28>;
			 resolution = <12>;
			 status = "okay";
		 };
		 testlabel2: ds18b202 {
			compatible = "maxim,ds18b20";
			family-code = <0x28>;
			resolution = <12>;
			status = "okay";
		};
	 };
```

then you can specify the rom for a given device like so:

```
struct sensor_value romval;
struct w1_rom w1rom;
w1_uint64_to_rom(2954138715793261711,&w1rom);
w1_rom_to_sensor_value(&w1rom,&romval);
sensor_attr_set(dev,0,SENSOR_ATTR_W1_ROM,&romval);
```

**Reply by dolence** [reply]
This is exactly what I was searching for. Thank you!
Now that I found this, I'm thinking it may not apply to my use case. I have 16 channels (two DS2482-800) and up to 30 sensors per channel. Would it mean I need to have 480 sensor entries on my dts file?

**Reply by CodingGhost** [reply]
@dolence no, the trick is to just put two sensors in the dts file, that enables the read by rom id stuff, regardless of how many you have.
I have 3 sensors (and still two dts entries) and perform sensor discovery on the first boot of the systemand save all roms. 
you could do the same.

**Reply by dolence** [reply]
That makes sense. I will give it a try, thank you!

**Reply by CodingGhost** [reply]
I just looked at my own code, and I do have 3 entries in my dts, but If I remember correctly I only did this because it made the code look nicer.
You should be able to just get one device and dynamically change the rom of that zephyr device to the rom of the actual device you want to read, then get the sensor value from that device.

Maybe I should make this a feature request, the current way of doing this seems very hacky..

**Reply by dolence** [reply]
Yeah, would be nice to multidrop and convert 30 or more sensors at once instead of read one by one. It goes against the logic of using a multichannel IC capable of this, like the ds2482-800 I'm using (and Zephyr has drivers for).

**Comment by dolence** [comment]
Thank you for everything. It's been hard to get help on this issue and I saw you have been through the same issues and not everyone share their findings. You are a nice person. Let me know if you open that FR, I have a huge interest on that subject.

**Reply by Martdur** [reply]
I'm facing the same issue right know, the solution seems legit but begin to be limited when you want to addressing to multiples devices.
I'll try to take some of the DS18B20 driver function and integrated in my project, that way I'll have control over timing and addressing or maybe did you find a better solution ?

**Reply by str4t0m** [reply]
An alternative to adding 2 device tree nodes is to enable the `CONFIG_W1_NET_FORCE_MULTIDROP_ADDRESSING=y` option which will also have the effect of using w1 match commands instead of the w1 skip commands.
Would that address your concerns or would need something different?

**Reply by Martdur** [reply]
What I want is addressing to multiples devices at one time to order them to make a conversion, then waiting for one conversion time and then asking the data to all device in a raw.
But when using the ds18b20 driver I have to wait for every single conversion when I ask for a fetch.

**Reply by dolence** [reply]
Did you find any solution to this?

**Reply by Martdur** [reply]
I've created a hal library in cpp that introduces W1Bus and W1Device classes, that way I can easily communicate as I want to one or to all devices using the hid driver.
I had to handle ds18b20 addresses and datas on my own but it works well.
I'm using AS6214ek btw.

**Comment by dolence** [comment]
> I've created a hal library in cpp that introduces W1Bus and W1Device classes, that way I can easily communicate as I want to one or to all devices using the hid driver. I had to handle ds18b20 addresses and datas on my own but it works well. I'm using AS6214ek btw.

I think I will have to do something similar. Interfacing too many sensors here.
